### PR TITLE
Allow host functions as table elements

### PR DESCRIPTION
**Pull request by rossberg-chromium**
As discussed here: https://github.com/WebAssembly/design/issues/798#issuecomment-245931553

Spec implementation: WebAssembly/spec#347

(Plus a few minor fixes, and updates to spec crossrefs.)

### CODE REVIEW

File: JS.md [hunk 1]
```diff
@@ -262,11 +262,6 @@ Set `moduleRecord.[[Namespace]]` to `moduleNamespace`.
 Let `instanceObject` be a new `WebAssembly.Instance` object setting
 `[[Instance]]` to `instance` and `exports` to `moduleNamespace`.
 
-If any of the elements of an [Elements section](Modules.md#elements-section)
-refers to an imported function which is not an
-[Exported Function Exotic Object](#exported-function-exotic-objects), throw a
-[`TypeError`](https://tc39.github.io/ecma262/#sec-native-error-types-used-in-this-standard-typeerror).
```
Comment (lukewagner) [inline JS.md]: There is a similar check above (GH won't give me a + to comment on it directly) preventing host functions in elems section that should be removed as well.

Comment (rossberg-chromium) [inline JS.md]: Ah, thanks. Removed.

**Comment by lukewagner** [comment]
Yes, sorry for the delay on this.  So two general comments:
* Because we're introducing new function identities here, I think we need to be explicit about where they are created and how they get propagated.  It seems like, in the initial pass through the import array, we need to coerce each element into some new array of coerced values that are referred to by exports and the elems section.
* By wholly removing the check on `set`, this allows arbitrary (non-webassembly-wrapped) JS functions which now don't have a destination signature and will require a totally new mechanism in the impl to handle in tables (esp. if the engine has an efficient indirect-call ABI).  I can see how it'd be useful, but could we keep `set` unchanged for now?

**Comment by rossberg-chromium** [comment]
Ah, good points. Reinstantiated the restriction on `set` and introduced explicit caching for host imports.

File: JS.md [hunk 2]
```diff
@@ -180,6 +180,8 @@ Let `imports` be an initially-empty [`import list`](https://github.com/WebAssemb
 * a [`Table.table`](#webassemblytable-objects)
 * a [`Values.value`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/values.ml#L9)
 
+Let `jafunctions` and `wrappers` be empty lists.
```
Comment (jfbastien) [inline JS.md]: "JS"

File: JS.md [hunk 3]
```diff
@@ -433,7 +434,7 @@ A `WebAssembly.Table` object contains a single [table](Semantics.md#table)
 which can be simultaneously referenced by multiple `Instance` objects. Each
 `Table` object has two internal slots:
  * [[Table]] : a [`Table.table`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/table.mli)
- * [[Values]] : an array whose elements are either `null` or Function Objects
+ * [[Values]] : an array whose elements are either `null` or callable objects
```
Comment (lukewagner) [inline JS.md]: Instead of "callable", could this be "Exported Function Exotic Objects"?

File: JS.md [hunk 4]
```diff
@@ -522,12 +520,8 @@ or `null`, throw a [`TypeError`](https://tc39.github.io/ecma262/#sec-native-erro
 
 Let `i` be the result of [`ToNonWrappingUint32`](#tononwrappinguint32)(`index`).
 
-If `v` is an [Exported Function Exotic Object](#exported-function-exotic-objects):
-* Set the `i`th element of `T.[[Table]]` to the `v.[[FunctionIndex]]`th function
-  in the module's [function index space](Modules.md#function-index-space).
```
Comment (rossberg-chromium) [inline JS.md]: Actually, looking at it now, I'm confused about the purpose of T.[[Table]], which doesn't seem to be used for anything. Also, the representation in the spec changed: a table no longer contains indices but actually a function values as well. Is this just essentially duplicating the same thing as T.[[Values]] now? Should I try to get rid of one?

Comment (rossberg-chromium) [inline JS.md]: Yes, but when we allowed tables to be im/exported the spec's representation had to change, and table elements are now of type `closure`, as defined here:
https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/instance.ml#L11
This is pretty much the spec's rendering of the [[Values]] array. The index is gone, we need to find some language saying how JS functions map to the spec's "host functions" to make the connection. And at that point we are essentially just duplicating the information, I think.

**Comment by rossberg-chromium** [comment]
This was *much* more work than I anticipated, but I think I managed to adapt the text to the current spec while defining the reuse of created objects more precisely.

File: JS.md [hunk 5]
```diff
@@ -173,83 +173,90 @@ is thrown. If the list of
...
-  * Otherwise, append `v.[[Table]]` to `imports`.
+  * Append `v` to `tables`.
+  * Append `v.[[Table]]` to `imports`.
 
 Let `instance` be the result of creating a new
-[`Eval.instance`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/eval.ml#L15)
-given `module` and `imports`. Note: this step does not mutate imported memories
-or tables.
+[`Eval.init`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/eval.ml#L416)
+given `module` and `imports`.
```
Comment (lukewagner) [inline JS.md]: How about "creating a new [Eval.instance] by calling [Eval.init] given `module` and `imports`"?

Comment (rossberg-chromium) [inline JS.md]: Good point, I actually misparsed the original sentence...

**Comment by lukewagner** [comment]
Thanks for all the work this is much clearer! lgtm

**Pull request merged**